### PR TITLE
Messenger

### PR DESCRIPTION
**Pull request by wkorol**
Injected working messenger into the project. Changed structure of choosing the provider.

### CODE REVIEW

File: src/Command/ImportCommand.php [hunk 1]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace App\Command;
```
Comment (GITFenix) [inline src/Command/ImportCommand.php:5]: please add strict_type declaration

File: src/Command/ImportCommand.php [hunk 2]
```diff
@@ -10,28 +11,33 @@
...
-        
+
         parent::__construct();
     }
 
-    public function checkProviderExistance($providerName) : bool {
-        if(in_array(strtolower($providerName), $this->providers)) {
+    public function checkProviderExistance($providerName): bool
+    {
+        if (in_array(strtolower($providerName), $this->providers)) {
```
Comment (l3l0) [inline src/Command/ImportCommand.php]: We should use strict comparation in `in_array` so thrid params in in_array to `true`

Please check this one:

```php
<?php

$arr = [''];
$value = 0;

var_dump(in_array($value, $arr));
var_dump(in_array($value, $arr, true));
```

File: src/Command/ImportCommand.php (hunk 1 above)
Comment (wkorol) [reply src/Command/ImportCommand.php:5]: https://github.com/wkorol/vcs-import-project/pull/15/commits/a6111d0946c042ccb29d99c5ff30717f33fedad6

File: src/Command/ImportCommand.php [hunk 3]
```diff
@@ -10,28 +11,33 @@
...
-    private $githubService;
     private $repoRepository;
-    private $bitbucketService;
     private $providers;
-    
-    public function __construct(GithubService $githubService, RepoRepository $repoRepository, bitbucketService $bitbucketService, ParameterBagInterface $params)
+    private $importCommandCreator;
+    private $bus;
+
+    public function __construct(MessageBusInterface $bus, ImportCommandCreator $importCommandCreator, RepoRepository $repoRepository, ParameterBagInterface $params)
```
Comment (GITFenix) [inline src/Command/ImportCommand.php]: I don't see $repoRepository usage so we can remove it from  constructor.
The same with params. We only use it to check if given provider is handled by us. Instead we could delegate it to our ImportCommandCreator. If non command was created let throw exception. I will comment it on the ImportCommandCreator

File: src/Command/ImportCommand.php [hunk 4]
```diff
@@ -10,28 +11,33 @@
...
+
         $this->providers = $params->get('providers');
-        
+
         parent::__construct();
     }
 
-    public function checkProviderExistance($providerName) : bool {
-        if(in_array(strtolower($providerName), $this->providers)) {
+    public function checkProviderExistance($providerName): bool
```
Comment (GITFenix) [inline src/Command/ImportCommand.php]: this method is no longer used

File: src/Command/ImportCommand.php [hunk 5]
```diff
@@ -44,33 +50,21 @@ protected function configure()
...
-        
-    }
-    
 
+
+        //$this->githubService->importToDb($orgName);
+
+        //$this->bitbucketService->importToDb($orgName);
+
+        $importCommand = $this->importCommandCreator->create($input->getArgument('username'), $input->getArgument('provider'));
```
Comment (GITFenix) [inline src/Command/ImportCommand.php]: we can use try / catch construct 
```suggestion
        try {
            $importCommand = $this->importCommandCreator->create($input->getArgument('username'), 
            $input->getArgument('provider'));
        } catch (ProviderNotFound $exceptioin) {
            $output->writeln($provider . 'is not implemented yet in our system.')
            return Command::FAILURE; 
        }
```

File: src/Message/ImportCommandCreator.php [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+class ImportCommandCreator {
+    public function create($username, $provider) {
+        if ($provider == 'github') {
+            return new GithubImportCommand($username);
+        }
+        if($provider == 'bitbucket') {
+            return new BitbucketImportCommand($username);
+        }
+    }
```
Comment (GITFenix) [inline src/Message/ImportCommandCreator.php:26]: now we can throw exception if there is no provider 
```suggestion
        throw new ProviderNotFound();
    }

```
and then we can catch this exception in the ImportCommand

File: src/MessageHandler/BitbucketImportCommandHandler.php [hunk 7]
```diff
@@ -0,0 +1,24 @@
+<?php
+declare(strict_types=1);
```
Comment (GITFenix) [inline src/MessageHandler/BitbucketImportCommandHandler.php:2]:
```suggestion

declare(strict_types=1);

```

File: src/Message/ImportCommandCreator.php [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+use App\MessageHandler\BitbucketImportCommandHandler;
+use App\MessageHandler\GithubImportCommandHandler;
+
+class ImportCommandCreator {
+    public function create($username, $provider) {
+        if ($provider == 'github') {
+            return new GithubImportCommand($username);
+        }
+        if($provider == 'bitbucket') {
```
Comment (GITFenix) [inline src/Message/ImportCommandCreator.php]:
```suggestion

        if($provider == 'bitbucket') {

```

File: src/Message/ImportCommandCreator.php [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+declare(strict_types=1);
+
+namespace App\Message;
+
+use App\MessageHandler\BitbucketImportCommandHandler;
+use App\MessageHandler\GithubImportCommandHandler;
+
+class ImportCommandCreator {
+    public function create($username, $provider) {
+        if ($provider == 'github') {
```
Comment (GITFenix) [inline src/Message/ImportCommandCreator.php]: better to compare strictly by '==='    thanks to this not only value is compared but also type

File: src/Message/ImportCommandCreator.php [hunk 10]
```diff
@@ -0,0 +1,18 @@
+<?php
+declare(strict_types=1);
+
+namespace App\Message;
+
+use App\MessageHandler\BitbucketImportCommandHandler;
+use App\MessageHandler\GithubImportCommandHandler;
+
+class ImportCommandCreator {
```
Comment (GITFenix) [inline src/Message/ImportCommandCreator.php]:
```suggestion
class ImportCommandCreator 
{

```

File: src/Message/ImportCommandCreator.php [hunk 11]
```diff
@@ -0,0 +1,18 @@
+<?php
+declare(strict_types=1);
+
+namespace App\Message;
+
+use App\MessageHandler\BitbucketImportCommandHandler;
+use App\MessageHandler\GithubImportCommandHandler;
+
+class ImportCommandCreator {
+    public function create($username, $provider) {
```
Comment (GITFenix) [inline src/Message/ImportCommandCreator.php]:
```suggestion
    public function create(string $username, string $provider) 
    {

```

File: src/Message/ImportCommandCreator.php [hunk 12]
```diff
@@ -0,0 +1,18 @@
+<?php
+declare(strict_types=1);
```
Comment (GITFenix) [inline src/Message/ImportCommandCreator.php:3]:
```suggestion

declare(strict_types=1);

```

File: src/MessageHandler/BitbucketImportCommandHandler.php [hunk 13]
```diff
@@ -0,0 +1,24 @@
+<?php
+declare(strict_types=1);
+
+namespace App\MessageHandler;
+use App\Message\BitbucketImportCommand;
+use App\Message\ImportCommandCreator;
+use App\Services\BitbucketService;
+use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
+
+class BitbucketImportCommandHandler implements MessageHandlerInterface {
```
Comment (GITFenix) [inline src/MessageHandler/BitbucketImportCommandHandler.php]:
```suggestion
class BitbucketImportCommandHandler implements MessageHandlerInterface 
{

```

File: src/MessageHandler/BitbucketImportCommandHandler.php [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+use App\Message\BitbucketImportCommand;
+use App\Message\ImportCommandCreator;
+use App\Services\BitbucketService;
+use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
+
+class BitbucketImportCommandHandler implements MessageHandlerInterface {
+    private BitbucketService $bitbucketService;
+
+    public function __construct(BitbucketService $bitbucketService) {
+
```
Comment (GITFenix) [inline src/MessageHandler/BitbucketImportCommandHandler.php]:
```suggestion
    public function __construct(BitbucketService $bitbucketService) 
    {
```

File: src/MessageHandler/BitbucketImportCommandHandler.php [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+        $this->bitbucketService = $bitbucketService;
+    }
+
+    public function __invoke(BitbucketImportCommand $command) {
+        $this->bitbucketService->importToDb($command->getOrgName());
+
+
```
Comment (GITFenix) [inline src/MessageHandler/BitbucketImportCommandHandler.php]:
```suggestion

```

File: src/MessageHandler/GithubImportCommandHandler.php [hunk 16]
```diff
@@ -0,0 +1,24 @@
+<?php
+declare(strict_types=1);
```
Comment (GITFenix) [inline src/MessageHandler/GithubImportCommandHandler.php:3]:
```suggestion

declare(strict_types=1);

```

File: src/MessageHandler/GithubImportCommandHandler.php [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    public function __construct(GithubService $githubService) {
+        $this->githubService = $githubService;
+    }
+    public function __invoke(GithubImportCommand $command) {
+        $this->githubService->importToDb($command->getOrgName());
+    }
+
+}
+
+
```
Comment (GITFenix) [inline src/MessageHandler/GithubImportCommandHandler.php]:
```suggestion
class GithubImportCommandHandler implements MessageHandlerInterface 
{
    private GithubService $githubService;

    public function __construct(GithubService $githubService) 
    {
        $this->githubService = $githubService;
    }
    public function __invoke(GithubImportCommand $command): void 
    {
        $this->githubService->importToDb($command->getOrgName());
    }
}
```
sprawdź czy z tym voidem jest ok bo nie mogę podejrzeć do MessageHandlerInterface

File: src/Message/ImportCommandCreator.php (hunk 8 above)
Comment (l3l0) [inline src/Message/ImportCommandCreator.php]:
```suggestion
        if($provider === 'bitbucket') {

```

File: src/Message/ImportCommandCreator.php (hunk 9 above)
Comment (l3l0) [inline src/Message/ImportCommandCreator.php]:
```suggestion
        if ($provider === 'github') {

```

File: src/Message/ImportCommandCreator.php (hunk 11 above)
Comment (l3l0) [inline src/Message/ImportCommandCreator.php]:
```suggestion
    public function create(string $username, string $provider): ImportCommand {
```

if we have abstract type for command.

File: src/Message/ImportCommandCreator.php [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+use App\MessageHandler\GithubImportCommandHandler;
+
+class ImportCommandCreator {
+    public function create($username, $provider) {
+        if ($provider == 'github') {
+            return new GithubImportCommand($username);
+        }
+        if($provider == 'bitbucket') {
+            return new BitbucketImportCommand($username);
+        }
```
Comment (l3l0) [inline src/Message/ImportCommandCreator.php:23]: There we can do exercise: How to implement each new provider without needed of if or switch here? As we use AbstractFactory it should be possible

File: src/Message/GithubImportCommand.php [hunk 19]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Message;
+
+class GithubImportCommand
```
Comment (l3l0) [inline src/Message/GithubImportCommand.php]:
```suggestion
class GithubImportCommand extends ImportCommand
```

File: src/Message/BitbucketImportCommand.php [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+    private string $username;
+
+    public function __construct(string $username)
+    {
+        $this->username = $username;
+    }
+    public function getOrgName(): string
+    {
+        return $this->username;
+    }
```
Comment (l3l0) [inline src/Message/BitbucketImportCommand.php:12]: This seems like duplication with GithubImportCommand any idea how to remove duplicated code there?

File: src/Message/BitbucketImportCommand.php [hunk 21]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Message;
+
+class BitbucketImportCommand
```
Comment (l3l0) [inline src/Message/BitbucketImportCommand.php]:
```suggestion
class BitbucketImportCommand extends ImportCommand
```

You of couse need to create new abstract class `ImportCommand` https://www.php.net/manual/en/language.oop5.abstract.php

File: src/MessageHandler/BitbucketImportCommandHandler.php [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+class BitbucketImportCommandHandler implements MessageHandlerInterface {
+    private BitbucketService $bitbucketService;
+
+    public function __construct(BitbucketService $bitbucketService) {
+
+        $this->bitbucketService = $bitbucketService;
+    }
+
+    public function __invoke(BitbucketImportCommand $command) {
+        $this->bitbucketService->importToDb($command->getOrgName());
```
Comment (l3l0) [inline src/MessageHandler/BitbucketImportCommandHandler.php]: In such case importToDb is not needed any more you can move whole code from that method into handler directly

File: src/MessageHandler/GithubImportCommandHandler.php [hunk 23]
```diff
@@ -0,0 +1,24 @@
...
+use Symfony\Component\Messenger\Handler\MessageHandlerInterface;
+
+class GithubImportCommandHandler implements MessageHandlerInterface {
+    private GithubService $githubService;
+
+    public function __construct(GithubService $githubService) {
+        $this->githubService = $githubService;
+    }
+    public function __invoke(GithubImportCommand $command) {
+        $this->githubService->importToDb($command->getOrgName());
```
Comment (l3l0) [inline src/MessageHandler/GithubImportCommandHandler.php]: In such case importToDb is not needed any more you can move whole code from that method into handler directly

File: src/Command/ImportCommand.php (hunk 3 above)
Comment (wkorol) [reply src/Command/ImportCommand.php]: Fixed

File: src/Command/ImportCommand.php (hunk 4 above)
Comment (wkorol) [reply src/Command/ImportCommand.php]: Fixed

File: src/Message/ImportCommandCreator.php (hunk 6 above)
Comment (wkorol) [reply src/Message/ImportCommandCreator.php:26]: Fixed

File: src/Message/ImportCommandCreator.php (hunk 9 above)
Comment (wkorol) [reply src/Message/ImportCommandCreator.php]: Fixed

File: src/Message/ImportCommandCreator.php (hunk 11 above)
Comment (wkorol) [reply src/Message/ImportCommandCreator.php]: fixed

File: src/Message/ImportCommandCreator.php (hunk 12 above)
Comment (wkorol) [reply src/Message/ImportCommandCreator.php:3]: Fixed

File: src/MessageHandler/BitbucketImportCommandHandler.php (hunk 14 above)
Comment (wkorol) [reply src/MessageHandler/BitbucketImportCommandHandler.php]: Fixed

File: src/MessageHandler/BitbucketImportCommandHandler.php (hunk 15 above)
Comment (wkorol) [reply src/MessageHandler/BitbucketImportCommandHandler.php]: Fixed

File: src/MessageHandler/GithubImportCommandHandler.php (hunk 16 above)
Comment (wkorol) [reply src/MessageHandler/GithubImportCommandHandler.php:3]: Fixed

File: src/MessageHandler/GithubImportCommandHandler.php (hunk 17 above)
Comment (wkorol) [reply src/MessageHandler/GithubImportCommandHandler.php]: Jest ok

File: src/Message/ImportCommandCreator.php (hunk 8 above)
Comment (wkorol) [reply src/Message/ImportCommandCreator.php]: Fixed

File: src/Message/ImportCommandCreator.php (hunk 9 above)
Comment (wkorol) [reply src/Message/ImportCommandCreator.php]: Fixed

File: src/Command/ImportCommand.php (hunk 2 above)
Comment (wkorol) [reply src/Command/ImportCommand.php]: It's no longer needed.

File: src/MessageHandler/BitbucketImportCommandHandler.php (hunk 13 above)
Comment (wkorol) [reply src/MessageHandler/BitbucketImportCommandHandler.php]: Fixed

File: src/Message/ImportCommandCreator.php (hunk 10 above)
Comment (wkorol) [reply src/Message/ImportCommandCreator.php]: Fixed

File: src/MessageHandler/BitbucketImportCommandHandler.php (hunk 7 above)
Comment (wkorol) [reply src/MessageHandler/BitbucketImportCommandHandler.php:2]: Fixed

File: src/Message/ImportCommandCreator.php (hunk 11 above)
Comment (wkorol) [reply src/Message/ImportCommandCreator.php]: Fixed

File: src/Message/ImportCommandCreator.php (hunk 18 above)
Comment (wkorol) [reply src/Message/ImportCommandCreator.php:23]: Fixed

File: src/Message/GithubImportCommand.php (hunk 19 above)
Comment (wkorol) [reply src/Message/GithubImportCommand.php]: Founded solution

File: src/Message/BitbucketImportCommand.php (hunk 20 above)
Comment (wkorol) [reply src/Message/BitbucketImportCommand.php:12]: Founded solution

File: src/Message/BitbucketImportCommand.php (hunk 21 above)
Comment (wkorol) [reply src/Message/BitbucketImportCommand.php]: Founded solution

File: src/MessageHandler/BitbucketImportCommandHandler.php (hunk 22 above)
Comment (wkorol) [reply src/MessageHandler/BitbucketImportCommandHandler.php]: Solved

File: src/MessageHandler/GithubImportCommandHandler.php (hunk 23 above)
Comment (wkorol) [reply src/MessageHandler/GithubImportCommandHandler.php]: Solved

**Pull request merged**